File: engine-go/engine/pointbuy.go
package engine

import "fmt"

// Compra de pontos (book p17, Tabela 1-1) — ports t20-data/src/point-buy.ts.
// All six attributes start at 0 with a 10-point budget; costs 1→1, 2→2, 3→4,
// 4→7; exactly ONE attribute may drop to −1 to refund 1 point. Racial mods apply
// after, outside the budget.
const (
	PointBuyBudget = 10
	pointBuyMin    = -1
	pointBuyMax    = 4
)

var pointBuyCosts = map[int]int{-1: -1, 0: 0, 1: 1, 2: 2, 3: 4, 4: 7}

// PointBuyStatus is the front-facing point-buy result: total spent (nil when any
// value is out of range, mirroring pointBuySpent's throw) + advisory warnings.
type PointBuyStatus struct {
	Spent    *int     `json:"spent"`
	Warnings []string `json:"warnings"`
}

// PointBuySpent totals the points a BASE attribute spread costs. ok=false when
// any value is outside −1..4 (pointBuySpent throws there; the front shows "—").
func PointBuySpent(attrs map[string]int) (int, bool) {
	spent := 0
	for _, k := range AttributeKeys {
		cost, has := pointBuyCosts[attrs[k]]
		if !has {
			return 0, false
		}
		spent += cost
	}
	return spent, true
}

// PointBuyWarnings ports pointBuyWarnings: advisory p17 validation (empty = legal).
// Out-of-range values are reported, not thrown, so a live-editing UI stays up.
func PointBuyWarnings(attrs map[string]int) []string {
	warnings := []string{}
	spent, reduced := 0, 0
	for _, k := range AttributeKeys {
		v := attrs[k]
		if v < pointBuyMin || v > pointBuyMax {
			warnings = append(warnings, fmt.Sprintf(
				"compra de pontos: %s = %d fora do intervalo [%d, %d]", k, v, pointBuyMin, pointBuyMax))
			continue
		}
		if v == -1 {
			reduced++
		}
		spent += pointBuyCosts[v]
	}
	if reduced > 1 {
		warnings = append(warnings, fmt.Sprintf(
			"compra de pontos: apenas UM atributo pode ser reduzido a −1 (p17), há %d", reduced))
	}
	if spent > PointBuyBudget {
		warnings = append(warnings, fmt.Sprintf(
			"compra de pontos: %d pontos gastos excedem o limite de %d", spent, PointBuyBudget))
	}
	return warnings
}

// PointBuyStatusFor bundles spent + warnings for the WASM boundary.
func PointBuyStatusFor(attrs map[string]int) PointBuyStatus {
	warnings := PointBuyWarnings(attrs)
	if spent, ok := PointBuySpent(attrs); ok {
		return PointBuyStatus{Spent: &spent, Warnings: warnings}
	}
	return PointBuyStatus{Spent: nil, Warnings: warnings}
}

File: frontend/src/shared/rules/point-buy.ts
import { ATTRIBUTE_KEYS } from '@/shared/api/attribute-keys'
import type { AttributeKey } from '@/shared/api/attribute-keys'
/**
 * Compra de pontos (book p17, Tabela 1-1): all six attributes start at 0 and
 * the player gets 10 points. Costs: 1→1, 2→2, 3→4, 4→7. Exactly ONE attribute
 * may be reduced to −1 to refund 1 point ("reduzir UM atributo para −1").
 * Point-buy range is −1..4 (racial mods apply after, outside the budget).
 */

export const POINT_BUY_BUDGET = 10
export const POINT_BUY_MIN = -1
export const POINT_BUY_MAX = 4

const COSTS = new Map<number, number>([
  [-1, -1],
  [0, 0],
  [1, 1],
  [2, 2],
  [3, 4],
  [4, 7],
])

/** Cost of one attribute value. Throws for values outside −1..4. */
export function pointBuyCost(value: number): number {
  const cost = COSTS.get(value)
  if (cost === undefined) {
    throw new Error(
      `pointBuyCost: value must be in [${POINT_BUY_MIN}, ${POINT_BUY_MAX}], got ${value}`,
    )
  }
  return cost
}

/** Total points spent by a full BASE attribute spread (pre-race). */
export function pointBuySpent(
  attrs: Readonly<Record<AttributeKey, number>>,
): number {
  return ATTRIBUTE_KEYS.reduce((sum, k) => sum + pointBuyCost(attrs[k] ?? 0), 0)
}

/**
 * Advisory validation of a spread against the p17 rules. Empty = legal.
 * Out-of-range values are reported (not thrown) so a live-editing UI can
 * show the problem without crashing.
 */
export function pointBuyWarnings(
  attrs: Readonly<Record<AttributeKey, number>>,
): string[] {
  const warnings: string[] = []
  let spent = 0
  let reduced = 0
  for (const k of ATTRIBUTE_KEYS) {
    const v = attrs[k] ?? 0
    if (v < POINT_BUY_MIN || v > POINT_BUY_MAX) {
      warnings.push(
        `compra de pontos: ${k} = ${v} fora do intervalo [${POINT_BUY_MIN}, ${POINT_BUY_MAX}]`,
      )
      continue
    }
    if (v === -1) reduced++
    spent += pointBuyCost(v)
  }
  if (reduced > 1) {
    warnings.push(
      `compra de pontos: apenas UM atributo pode ser reduzido a −1 (p17), há ${reduced}`,
    )
  }
  if (spent > POINT_BUY_BUDGET) {
    warnings.push(
      `compra de pontos: ${spent} pontos gastos excedem o limite de ${POINT_BUY_BUDGET}`,
    )
  }
  return warnings
}
